infer/internal/ende/ende_test.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2022, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ende
+
+import (
+	"reflect"
+	"testing"
+
+	r "github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"pgregory.net/rapid"
+
+	rType "github.com/pulumi/pulumi-go-provider/internal/rapid/reflect"
+	rResource "github.com/pulumi/pulumi-go-provider/internal/rapid/resource"
+)
+
+// testRoundTrip asserts that the result of pMap can be decoded onto T, and then lossly
+// encoded back into a property map.
+func testRoundTrip[T any](t *testing.T, pMap func() r.PropertyMap) {
+	t.Run("", func(t *testing.T) {
+		t.Parallel()
+		var typeInfo T
+		toDecode := pMap()
+		encoder, err := Decode(toDecode, &typeInfo)
+		require.NoError(t, err)
+
+		assert.Equalf(t, pMap(), toDecode, "mutated decode map")
+
+		reEncoded, err := encoder.Encode(typeInfo)
+		require.NoError(t, err)
+		assert.Equal(t, pMap(), reEncoded)
+	})
+}
+
+func TestRapidRoundTrip(t *testing.T) {
+	t.Parallel()
+	rapid.Check(t, func(t *rapid.T) {
+		typed := rResource.ValueOf(rType.Struct(5)).Draw(t, "top-level")
+		pMap := func() r.PropertyMap { return typed.Value.ObjectValue().Copy() }
+		goValue := reflect.New(typed.Type).Interface()
+
+		toDecode := pMap()
+		encoder, err := decode(toDecode, goValue,
+			false /*ignoreUnrecognized*/, false /*allowMissing*/)
```

Comment:
why do you set ignoreUnrecognized & allowMissing to false here but not in testRoundTrip

Reply:
`testRoundTrip` uses the function that we expose to others: `ende.Decode`:

https://github.com/pulumi/pulumi-go-provider/blob/7484a3ff9bff06a7855806451e4e237d3656adda/infer/internal/ende/ende.go#L36-L38

Because it is generic over `T`, I can't use it in rapid testing, so I drop down to the lower level `decode` (with it's extra options).